dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.h

```diff
@@ -0,0 +1,107 @@
+
+    bool take(TaskPtr & task) override;
+
+    void updateStatistics(const TaskPtr & task, ExecTaskStatus, size_t inc_value) override;
+
+    bool empty() const override;
+
+    void finish() override;
+
+    void cancel(const String & query_id, const String & resource_group_name) override;
+
+private:
+    bool tryTakeCancelTaskWithoutLock(TaskPtr & task);
+
+    // <resource_group_name, resource_group_task_queues>
+    using ResourceGroupTaskQueue = std::unordered_map<std::string, std::shared_ptr<NestedQueueType>>;
+
+    // <priority, corresponding_task_queue, resource_group_name>
+    using ResourceGroupInfo = std::tuple<UInt64, std::shared_ptr<NestedQueueType>, std::string>;
+    using CompatorType = bool (*)(const ResourceGroupInfo &, const ResourceGroupInfo &);
+    using ResourceGroupInfoQueue = std::priority_queue<ResourceGroupInfo, std::vector<ResourceGroupInfo>, CompatorType>;
+
+    // Index of ResourceGroupInfo.
+    static constexpr auto InfoIndexPriority = 0;
+    static constexpr auto InfoIndexPipelineTaskQueue = 1;
+    static constexpr auto InfoIndexResourceGroupName = 2;
+    static constexpr auto DEFAULT_WAIT_INTERVAL_WHEN_RUN_OUT_OF_RU = std::chrono::seconds(1);
+
+    // ResourceGroupInfoQueue compator.
+    static bool compareResourceInfo(const ResourceGroupInfo & info1, const ResourceGroupInfo & info2)
+    {
+        auto priority1 = std::get<InfoIndexPriority>(info1);
+        auto priority2 = std::get<InfoIndexPriority>(info2);
+
+        // Return true means lower priority.
+        // Here we want make zero priority to be lower priority than positive priority.
+        // Because negative priority means corresponding resource group has no more token.
+        if (priority1 <= 0)
+            return true;
+        if (priority2 <= 0)
+            return false;
+        return priority1 > priority2;
+    }
+
+    // 1. Update cpu time/RU of resource group.
+    // 2. Update resource_group_infos and reorder resource_group_infos by priority.
+    void updateResourceGroupStatisticWithoutLock(const std::string & name, UInt64 consumed_cpu_time);
+
+    // Update resource_group_infos, will reorder resource group by priority.
+    void updateResourceGroupInfosWithoutLock();
+
+    // Submit task into task queue of specific resource group.
+    void submitWithoutLock(TaskPtr && task);
+
+    mutable std::mutex mu;
+    std::condition_variable cv;
+
+    bool is_finished = false;
+
+    CompatorType compator = compareResourceInfo;
+    ResourceGroupInfoQueue resource_group_infos{compator};
+    ResourceGroupTaskQueue resource_group_task_queues;
+
+    // <resource_group_name, acculumated_cpu_time>
+    // when acculumated_cpu_time >= YIELD_MAX_TIME_SPENT_NS, will update resource group.
+    // This is to prevent the resource group from being updated too frequently.
+    std::unordered_map<std::string, UInt64> resource_group_statistic;
+
+    std::unordered_map<std::string, std::shared_ptr<NestedQueueType>> cancel_query_ids;
```

Comment:
Maybe it would be simpler to maintain a cancel_queue in `ResourceControlQueue` like MLFQ.
```
bool take(TaskPtr & task) override
{
    if (!cancel_queue.empty())
        return popTask(cancel_queue);
}

void cancel(const String & query_id) override
{
    for (const auto & queue : resource_group_task_queues)
    {
        queue.moveCancelledTasks(query_id, cancel_queue);
    }
}
```
The time complexity of cancel is O(n), same as MLFQ, n is the total number of tasks.
This greatly simplifies the code of ResourceControlQueue.

Reply:
done